File: MathCore.DSP/Filters/ButterworthBandPass.cs
using System.Diagnostics;

using static System.Math;

using static MathCore.Consts;
using static MathCore.Polynom.Array;

// ReSharper disable InconsistentNaming
// ReSharper disable MemberCanBePrivate.Global
// ReSharper disable UnusedAutoPropertyAccessor.Global

namespace MathCore.DSP.Filters;

public class ButterworthBandPass : ButterworthFilter
{
    private static void CheckFrequencies(double dt, double fsl, double fpl, double fph, double fsh)
    {
        if (dt <= 0)
            throw new InvalidOperationException($"Период дискретизации dt={dt} не может быть меньше, либо равен нулю")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fsl), fsl },
                    { nameof(fpl), fpl },
                    { nameof(fph), fph },
                    { nameof(fsh), fsh },
                }
            };

        if (1 / dt == 0)
            throw new InvalidOperationException("Частота дискретизации не может быть равна нулю")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fsl), fsl },
                    { nameof(fpl), fpl },
                    { nameof(fph), fph },
                    { nameof(fsh), fsh },
                }
            };

        if (fsl >= fpl)
            throw new InvalidOperationException($"Нижняя частота среза fsl должна быть ниже нижней частоты пропускания fpl\r\n  dt={dt}\r\n  fsl={fsl}\r\n  fpl={fpl}\r\n  fph={fph}\r\n  fsh={fsh}")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fsl), fsl },
                    { nameof(fpl), fpl },
                    { nameof(fph), fph },
                    { nameof(fsh), fsh },
                }
            };

        if (fpl >= fph)
            throw new InvalidOperationException($"Нижняя частота пропускания fpl должна быть ниже верхней частоты пропускания fph\r\n  dt={dt}\r\n  fsl={fsl}\r\n  fpl={fpl}\r\n  fph={fph}\r\n  fsh={fsh}")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fsl), fsl },
                    { nameof(fpl), fpl },
                    { nameof(fph), fph },
                    { nameof(fsh), fsh },
                }
            };

        if (fph >= fsh)
            throw new InvalidOperationException($"Верхняя частота пропускания fph должна быть ниже верхней частоты среза fsh\r\n  dt={dt}\r\n  fsl={fsl}\r\n  fpl={fpl}\r\n  fph={fph}\r\n  fsh={fsh}")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fsl), fsl },
                    { nameof(fpl), fpl },
                    { nameof(fph), fph },
                    { nameof(fsh), fsh },
                }
            };

        if (fsh >= 1 / dt / 2)
            throw new InvalidOperationException($"Верхняя частота среза fsh должна быть ниже половины частоты дискретизации fd={1 / dt} (1 / (dt={dt}))\r\n  dt={dt}\r\n  fsl={fsl}\r\n  fpl={fpl}\r\n  fph={fph}\r\n  fsh={fsh}")
            {
                Data =
                {
                    { nameof(dt), dt },
                    { "fd", 1/dt },
                    { nameof(fsl), fsl },
                    { nameof(fpl), fpl },
                    { nameof(fph), fph },
                    { nameof(fsh), fsh },
                }
            };
    }

    /// <summary>Формирование спецификации фильтра</summary>
    /// <param name="dt">Период дискретизации сигнала</param>
    /// <param name="fsl">Частота нижней границы полосы заграждения</param>
    /// <param name="fpl">Частота нижней границы полосы пропускания</param>
    /// <param name="fph">Частота верхней границы полосы пропускания</param>
    /// <param name="fsh">Частота верхней границы полосы заграждения</param>
    /// <param name="Gp">Уровень АЧХ в полосе пропускания (в единицах 0..1)</param>
    /// <param name="Gs">Неравномерность АЧХ в полосе заграждения (в единицах 0..1)</param>
    /// <returns>Спецификация фильтра</returns>
    private static Specification GetSpecification(
        double dt,
        double fsl,
        double fpl,
        double fph,
        double fsh,
        double Gp = 0.891250938,
        double Gs = 0.031622777)
    {
        CheckFrequencies(dt, fsl, fpl, fph, fsh);

        if (Gp <= Gs)
            throw new ArgumentOutOfRangeException(
                nameof(Gp), Gp, $"Уровень АЧХ в полосе пропускания Gp={Gp} был меньше, либо равен уровню АЧХ в полосе заграждения Gs={Gs}");

        var Fsl = ToDigitalFrequency(fsl, dt);
        var Fpl = ToDigitalFrequency(fpl, dt);
        var Fph = ToDigitalFrequency(fph, dt);
        var Fsh = ToDigitalFrequency(fsh, dt);

        var Wsl = pi2 * Fsl;
        var Wpl = pi2 * Fpl;
        var Wph = pi2 * Fph;
        var Wsh = pi2 * Fsh;

        var Wc = Wpl * Wph;
        var dW = Wph - Wpl;

        // Выбор опорной частоты
        // Если   Wc / Wsh > Wsl
        // то есть      Wc > Wsl*Wsh
        // то есть Wsl*Wsh > Wsl*Wsh
        // то есть центральная частота по границам подавления > центральной частоты по границам пропускания
        // то выбираем в качестве опорной частоты выбираем верхнюю границу пропускания
        // иначе, выбираем нижнюю границу пропускания
        var Ws = Wc / Wsh > Wsl
            ? Wsh
            : Wsl;
        var Fp = Abs(dW * Ws / (Wc - Ws.Pow2())) / pi2;
        const double Fs = 1 / pi2;

        // Для передачи информации о граничных частотах в спецификацию аналогвого прототипа перечситываем частоты цифрового фильтра обратно
        var fp = ToAnalogFrequency(Fp, dt);
        var fs = ToAnalogFrequency(Fs, dt);

        return new Specification(dt, fp, fs, Gp, Gs);
    }

    private static (double[] A, double[] B) Initialize(double fpl, double fph, Specification Spec)
    {
        // Пересчитываем аналоговые частоты полосы заграждения в цифровые
        var dt = Spec.dt;

        var Wpl = pi2 * ToDigitalFrequency(fpl, dt);
        var Wph = pi2 * ToDigitalFrequency(fph, dt);

        var dW = Wph - Wpl;

        var N = (int)Ceiling(Log(Spec.kEps) / Log(Spec.kW));
        Debug.Assert(N > 0, $"N > 0 :: {N} > 0");
        var poles = GetNormPoles(N, Spec.EpsP);

        var ppf_poles = TransformToBandPassW(poles, Wpl, Wph).ToArray();

        var ppf_zeros_z = Enumerable
           .Repeat(Complex.ReValue(1), N)
           .AppendLast(Enumerable.Repeat(Complex.ReValue(-1), N))
           .ToArray();
        var ppf_poles_z = ToZArray(ppf_poles, Spec.dt);

        var g_norm = Complex.Real;
        var fd2 = 2 * Spec.fd;
        for (var i = 0; i < N; i++)
        {
            var (p1, p2) = (ppf_poles[2 * i], ppf_poles[2 * i + 1]);
            var p = (fd2 - p1) * (fd2 - p2);
            g_norm *= fd2 * dW / p;
        }

        var G_norm = g_norm / Spec.EpsP;

        var B = GetCoefficientsInverted(ppf_zeros_z).ToArray(v => (v * G_norm).Re);
        var A = GetCoefficientsInverted(ppf_poles_z).ToRe();

        return (A, B);
    }

    public ButterworthBandPass(
        double dt,
        double fsl,
        double fpl,
        double fph,
        double fsh,
        double Gp = 0.891250938,
        double Gs = 0.01)
        : this(fsl, fpl, fph, fsh, GetSpecification(dt, fsl, fpl, fph, fsh, Gp, Gs))
    { }

    private ButterworthBandPass(double fsl, double fpl, double fph, double fsh, Specification Spec)
        : this(Initialize(fpl, fph, Spec), Spec) { }

    private ButterworthBandPass((double[] A, double[] B) Polynoms, Specification Spec) : this(Polynoms.B, Polynoms.A, Spec) { }

    private ButterworthBandPass(double[] B, double[] A, Specification Spec) : base(B, A, Spec) { }
}

File: MathCore.DSP/Filters/Python/filters/butterworth_bandpass.py
# Полосовой фильтр Баттерворта (порт C# ButterworthBandPass)
import numpy as np
from math import ceil, log, pi, tan, atan, sqrt, fabs

from .butterworth_filter import ButterworthFilter
from .analog_based_filter import AnalogBasedFilter, Specification
from .digital_filter import DigitalFilter

pi2 = 2.0 * pi


class ButterworthBandPass(ButterworthFilter):
    """
    Цифровой полосовой фильтр Баттерворта (ППФ).

    Нули в z-области: N нулей в точке z=+1 (постоянная составляющая)
    и N нулей в точке z=-1 (частота Найквиста).
    """

    @staticmethod
    def _get_specification(dt: float, fsl: float, fpl: float, fph: float, fsh: float,
                           Gp: float, Gs: float) -> Specification:
        """Формирует спецификацию ППФ через эквивалентный ФНЧ-прототип."""
        Fsl = DigitalFilter.to_digital_frequency(fsl, dt)
        Fpl = DigitalFilter.to_digital_frequency(fpl, dt)
        Fph = DigitalFilter.to_digital_frequency(fph, dt)
        Fsh = DigitalFilter.to_digital_frequency(fsh, dt)

        Wsl, Wpl = pi2 * Fsl, pi2 * Fpl
        Wph, Wsh = pi2 * Fph, pi2 * Fsh

        Wc = Wpl * Wph   # квадрат центральной частоты
        dW = Wph - Wpl   # ширина полосы пропускания

        # выбираем опорную частоту: верхнюю или нижнюю границу заграждения
        Ws = Wsh if (Wc / Wsh > Wsl) else Wsl
        Fp = fabs(dW * Ws / (Wc - Ws ** 2)) / pi2
        Fs = 1.0 / pi2  # нормировка прототипа

        fp = DigitalFilter.to_analog_frequency(Fp, dt)
        fs = DigitalFilter.to_analog_frequency(Fs, dt)
        return Specification(dt, fp, fs, Gp, Gs)

    @staticmethod
    def _initialize(fpl: float, fph: float, spec: Specification):
        """
        Расчёт коэффициентов ППФ Баттерворта.

        Алгоритм
        --------
        1. N = ceil(log(kEps)/log(kW)).
        2. Нормированные полюса ФНЧ-прототипа.
        3. LP→BP трансформация: каждый полюс даёт 2 полюса ППФ.
        4. Нули в z-области: N×(z=+1) + N×(z=-1).
        5. Нормировка усиления по аналоговым ВЧ-полюсам:
           g = ∏ fd2·dW / ((fd2-p1)(fd2-p2)) / EpsP.
        """
        dt = spec.dt
        Wpl = pi2 * DigitalFilter.to_digital_frequency(fpl, dt)
        Wph = pi2 * DigitalFilter.to_digital_frequency(fph, dt)
        dW = Wph - Wpl

        N = int(ceil(log(spec.kEps) / log(spec.kW)))

        poles = ButterworthFilter.get_norm_poles(N, spec.EpsP)

        # LP→BP трансформация (каждый LP-полюс → 2 BP-полюса)
        ppf_poles = AnalogBasedFilter.transform_to_band_pass_w(poles, Wpl, Wph)

        # нули в z-области: N штук z=+1 и N штук z=-1
        z_zeros = np.array([1.0 + 0j] * N + [-1.0 + 0j] * N)

        # билинейное z-преобразование аналоговых BP-полюсов
        z_poles = DigitalFilter.to_z_array(ppf_poles, dt)

        # нормировка усиления по парам аналоговых BP-полюсов
        fd2 = 2.0 * spec.fd
        g_norm = complex(1.0, 0.0)
        for i in range(N):
            p1, p2 = ppf_poles[2 * i], ppf_poles[2 * i + 1]
            g_norm *= fd2 * dW / ((fd2 - p1) * (fd2 - p2))
        G_norm = g_norm / spec.EpsP

        B = np.poly(z_zeros).real * G_norm.real
        A = np.poly(z_poles).real

        return A, B

    def __init__(self, dt: float, fsl: float, fpl: float, fph: float, fsh: float,
                 Gp: float = 0.891250938, Gs: float = 0.01):
        """
        Создаёт ППФ Баттерворта.

        Параметры
        ---------
        dt  : период дискретизации (с)
        fsl : нижняя граница полосы заграждения (Гц)
        fpl : нижняя граница полосы пропускания (Гц)
        fph : верхняя граница полосы пропускания (Гц)
        fsh : верхняя граница полосы заграждения (Гц)
        Gp  : коэффициент передачи на границе пропускания
        Gs  : коэффициент передачи на границе заграждения
        """
        spec = ButterworthBandPass._get_specification(dt, fsl, fpl, fph, fsh, Gp, Gs)
        A, B = ButterworthBandPass._initialize(fpl, fph, spec)
        super().__init__(B, A, spec)
